### DISCUSSION TITLE
Keycloak as a OIDC provider

### DISCUSSION THREAD

**Discussion by Colin-988** [rustdesk/rustdesk-server-pro#292] [General]
How can I provision Keycloak using OpenID Connect (OIDC) for secure authentication, and what is the correct syntax to specify a custom certificate in the PEM file format for Keycloak configuration?
![grafik](https://github.com/rustdesk/rustdesk-server-pro/assets/175134048/3e05cf2d-b3dc-487f-b561-194da25bc7c2)

**Comment by fufesou** [comment]
Hi, what OS do you use as the server?

https://github.com/rustdesk/rustdesk-server-pro/issues/130#issuecomment-1763392397

**Comment by Colin-988** [comment]
Linux Host with Docker Container

**Reply by fufesou** [reply]
You can add the docker volume mapping, and set the path in docker.

**Comment by fufesou** [comment]
Hello, is the configuration ok?
Please post here if there are any problems.

**Comment by Colin-988** [comment]
Im new to Docker, could you help maybe help to map the volume in docker. i didnt started my container with docker compose.

**Reply by rustdesk** [reply]
https://docs.docker.com/storage/volumes/

**Comment by Colin-988** [comment]
is the syntax right?
![grafik](https://github.com/rustdesk/rustdesk-server-pro/assets/175134048/f1315a87-9cb5-426b-8c7b-8f28044aa1fb)

**Reply by fufesou** [reply]
Hi, can you find where `db.sqlite3` is? You can find it by running `find data -name *.sqlite3`.
You can put `xxx.pem` in the same directory of `db.sqlite3`.

Then you can use `xxx.pem` in the configuration input box.

**Comment by Colin-988** [comment]
still the same error, i placed my pem file into the directory where the db is located. (permissions of the file are the same like the db)

**Reply by fufesou** [reply]
Can you please try `/root/xxx.pem` in the input box?

**Comment by Colin-988** [comment]
its also not wokring, i checked the bash console of the container but the new pem file is not in the directory of the conatiner. I tryed to restart the container  too

**Reply by fufesou** [reply]
Can you see `*.sqlite3` in the directory of the container?

**Reply by fufesou** [reply]
Can you please show me how you run the docker container? thanks.

**Comment by Colin-988** [comment]
now i just copied the fie directly in the directory, now the certificate problem should be fixed. still getting an error but my container logs show no error. 
![grafik](https://github.com/rustdesk/rustdesk-server-pro/assets/175134048/41f247f4-e07c-44a0-9243-a8b1634b3750)

**Reply by fufesou** [reply]
Please give full screenshot of the error, you can smudge out the url.

Please also check the log, it should record a more detailed error.

**Comment by Colin-988** [comment]
but could be an Keycloak problem, i check

**Comment by Colin-988** [comment]
Im currently running my web management console on http, its maybe also problem. Can you help me to setup https for that?

**Comment by rustdesk** [comment]
Here is tutorial of https setup: https://rustdesk.com/docs/en/self-host/rustdesk-server-pro/faq/#set-up-https-for-web-console-manually

**Comment by Colin-988** [comment]
I had to redeploy my container, now i had to setup everything. Is there any way to change the Private Key?

**Reply by rustdesk** [reply]
You can change it on web console, i guess it is not hard to find it for normal users.

**Comment by Colin-988** [comment]
sry i mean the public Key

**Reply by rustdesk** [reply]
Change private key, it will change public key. You can not change public key alone. It is part of private key.

**Comment by Colin-988** [comment]
So with the "old" Container i deployed some custom clients. Now when i try to connect it gives an connection Error (Keys mismatches)

**Reply by rustdesk** [reply]
https://github.com/rustdesk/rustdesk-server-pro/discussions/184

**Comment by Colin-988** [comment]
i dont have a backup from the old containers :D, is there any other possibility?

**Reply by rustdesk** [reply]
BTW, the data files (the private key file `id_ed25519` etc in https://github.com/rustdesk/rustdesk-server-pro/discussions/184) are not in the docker container if you follow our docker command in doc, it is on your local drive, then mapping to docker volume (https://docs.docker.com/storage/volumes/).

e.g. `sudo docker run --name hbbs -v ./data:/root -td --net=host --restart unless-stopped rustdesk/rustdesk-server-pro hbbs`

`-v ./data:/root` is a volume mapping, `./data` is the directory on your host, `/root` is the docker volume. `hbbs` writes all data files on `/root`, so you can see all data files under `./data` on your host.

If you do not specify `-v ./data:/root`, you will do lose all data files after container killed.

**Reply by rustdesk** [reply]
ChatGPT is good at such thing.

![image](https://github.com/rustdesk/rustdesk-server-pro/assets/71636191/2aa6ac14-902a-4065-8615-2817477d93c9)

**Comment by Colin-988** [comment]
I mean im still on the same server but didnt change the Key when i deploy the container?

**Reply by rustdesk** [reply]
If you lost the old `private key`, there is no way to restore it, we can not do magic, :(. we can get `public key` from `private key`, but no way to get private key from `public key`, because `public key` is a part of `private key`. You have to create new custom client on the new web console with new key.

If you have the old private key, you can paste it on new web console, or copy the `id_ed25519` (private key) file to new working directory (where you can see `db.sqlite3`).

**Comment by Colin-988** [comment]
thx for the help  #close
